### ISSUE TITLE
Feature Request: Request.copy(headers) - to replace all headers

### ISSUE THREAD

**Issue by bastman** [http4k/http4k#377]
Is there a simple way, to copy a Request and replace all headers?

In case of MemoryRequest this works well:
```
val req:MemoryRequest(...).copy(headers=newHeaders)
```
In case of RoutedRequest, I could not find a simple solution.

```
data class MemoryRequest(...  override val headers: Headers = listOf()) {
}

data class RoutedRequest(...) {
  // append headers
   override fun headers(headers: Headers): Request
}


```

Thanks a lot for your help :)

**Comment by daviddenton** [comment]
There's no real need to this in the lib as this should work fine: ```    val initial = Request(GET, "/first")
    Request(GET, "/second").headers(initial.headers)
```

**Comment by daviddenton** [comment]
On further inspection, you could do this simply to replace all the headers:

```
    fun Request.replaceHeaders(other: Request) = headers.fold(this) { acc, next ->
        acc.removeHeader(next.first)
    }.headers(other.headers)

    val initial = Request(GET, "/first")
    Request(GET, "/second").replaceHeaders(initial)
```

**Comment by bastman** [comment]
@daviddenton , I guess I did not make clear what feature I meant. I'll try again ...

```
request.headers(headers:newHeaders)

would behave different to 

request.copy(headers:newHeaders) // if that function  would exist
```
```
MemoryRequest(...  override val headers)

val req = MemoryRequest(...)
.copy(headers=newHeaders) . // overrides all headers

val req = RoutedRequest(...)
.headers(additionalHeaders) // adds these headers
```


I would like to be able to overrides all headers as shown in MemoryRequest.
Currently there is no simple way in case of RoutedRequest.

Therefore I have to create a "work-a-round", which feels kind of ugly to me ...

```
fun Request.withHeaders(headers: Headers) = when(this) {
    is MemoryRequest -> copy(headers = headers)
    is RoutedRequest -> withoutHeaders().addHeaders(headers)
    else-> error("Request.withHeaders() not implemented for clazz ${this::class.qualifiedName}")
}

fun Request.withoutHeaders(): Request = when(this) {
    is MemoryRequest -> copy(headers = listOf())
    is RoutedRequest -> {
        val src:Request = this
        val headerNames = src.headers.map { it.first }
        val sink = headerNames.fold(src) {req,headerName:String->
            req.removeHeader(headerName)
        }
        if(!sink.headers.isEmpty()) {
            error("Request.withoutHeaders(): failed to remove all headers!")
        }
        sink
    }
    else-> error("Request.withHeaders() not implemented for clazz ${this::class.qualifiedName}")
}



```

**Comment by daviddenton** [comment]
This is fixed now - although the signature is ```fun replaceHeaders(source: Headers): HttpMessage```. It will be released in 3.237.0 in the near future.
